### ISSUE TITLE
Deprecate environment variables which do not interoperate with file config when ready for stabilization

### ISSUE THREAD

**Issue by jack-berg** [open-telemetry/opentelemetry-specification#3967]
As discussed [here](https://github.com/open-telemetry/opentelemetry-specification/pull/3948#issuecomment-2025825640), we do not want to recommend multiple competing configuration stores. When file configuration is ready for stabilization, we should deprecate any [environment variables](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/configuration/sdk-environment-variables.md) which do not interoperate.

**Comment by VidhiPhatnani** [comment]
Issue: When attempting to use both the otel-profiling JAR (for trace-to-profiling linking) and the opentelemetry-samplers JAR (specified via OTEL_JAVAAGENT_EXTENSIONS env variable), the Opentelemetry Java Agent does not handle them together.

- If only otel-profiling jar is set as an environment variable, profiling and linking work correctly.
- If both the otel-profiling jar as env variable as well as opentelemetry-samplers jar as jvm arg ( -Dotel.javaagent.extensions ) are set then the trace-to-profiling linking stops working, but the SDK configuration is correctly applied.

Additionally, the sdk-config.yaml does not seem to support any other jar when provided as an environment variable.

I would really appreciate any guidance on resolving this issue. Thank you for the insights.

**Comment by jack-berg** [comment]
I'm not familiar with the otel-profiling jar - can you link to the source code for that?

> Additionally, the sdk-config.yaml does not seem to support any other jar when provided as an environment variable.

Can you elaborate?

**Comment by VidhiPhatnani** [comment]
The reference to the source code of otel-profiling jar : https://github.com/grafana/otel-profiling-java/tree/main (This dependency is an extension of opentelemetry-javaagent for providing link between traces and profiles)

1. **Working configuration (Only otel-profiling jar enabled using env variable)**
   export OTEL_JAVAAGENT_EXTENSIONS=<path to otel-profiling jar>
   jar -javaagent:<path to opentelemetry-javaagent jar> -jar myapp.jar

  This configuration enables the extension and the trace to profile linking is seen.

2. **Samplers and Otel-profiling jar added**
   export OTEL_JAVAAGENT_EXTENSIONS=<path to otel-profiling jar>
   jar -javaagent:<path to opentelemetry-javaagent jar> -Dotel.javaagent.extensions=<path to opentelemetry-samplers jar> -Dotel.experimental.config.file=<path to sdk config yml>
 
 This configuration works fine for sdk config but the trace to profile linking does not seem to work. 
 
**Expected behavior :** 
 
-   Both the otel-profiling and opentelemetry-samplers jars should work together without disabling the trace-to-profile linking.
-   sdk-config.yaml should support otel-profiling jar when set via environment variable.

Is there a recommended way to configure both profiling and sdk extensions together ? 

Thanks for the help!

**Comment by VidhiPhatnani** [comment]
For the config (The path is not visible above)
1. **Working configuration (Only otel-profiling jar enabled using env variable)**
export OTEL_JAVAAGENT_EXTENSIONS="/path to otel-profiling jar"
jar -javaagent:"/path to opentelemetry-javagent jar"  -jar myapp.jar

This configuration enables the extension and the trace to profile linking is seen.

2. **Samplers and Otel-profiling jar added**
export OTEL_JAVAAGENT_EXTENSIONS="/path to otel-profiling jar"
jar -javaagent:"/path to opentelemetry-javaagent jar" -Dotel.javaagent.extensions="/path to opentelemetry-samplers jar" -Dotel.experimental.config.file="/path to sdk -config yml file"

This configuration works fine for sdk config but the trace to profile linking does not seem to work.

**Comment by jack-berg** [comment]
> sdk-config.yaml should support otel-profiling jar when set via environment variable.

I agree. But this isn't the right repository to discuss that. This issue is about the interop story between the environment variables [defined in the spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/configuration/sdk-environment-variables.md) and declarative config. The spec doesn't dictate the behavior of `OTEL_JAVAAGENT_EXTENSIONS` or the `otel-profiling.jar`. Can you open an issue in https://github.com/grafana/otel-profiling-java/tree/main?

We may also need to make changes to https://github.com/open-telemetry/opentelemetry-java-instrumentation, but let's start with `grafana/otel-profiling-java` and see where that conversation leads.

**Comment by pellared** [comment]
> we should deprecate any [environment variables](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/configuration/sdk-environment-variables.md) which do not interoperate.

Do you have any examples?

**Comment by jack-berg** [comment]
We "sort of" maintain a list here: https://github.com/open-telemetry/opentelemetry-configuration/blob/b68188bf987f9b647ce4b190d3b757828d224983/examples/sdk-migration-config.yaml#L10-L33

I say sort of, because several of those could be used in declarative config but just aren't referenced in the `sdk-migration-config.yaml` file for various reasons. E.g. `OTEL_EXPORTER_ZIPKIN_ENDPOINT` is perfectly valid to reference, but we export to OTLP by default so there's no place for zipkin env vars in the default config.

Here's a filtered version of the list of those which really don't have any mapping because of core differences in how the information is modeled:

- OTEL_TRACES_SAMPLER
- OTEL_TRACES_SAMPLER_ARG
- OTEL_TRACES_EXPORTER
- OTEL_METRICS_EXPORTER
- OTEL_LOGS_EXPORTER
- OTEL_EXPORTER_OTLP_PROTOCOL, OTEL_EXPORTER_OTLP_{SIGNAL}_PROTOCOL

**Comment by pellared** [comment]
What does deprecating mean in this scenario? 

That we will frezee this environment variables in the specification?
Do we expect languages to also give some deprecation notice?
Do we want to eventually completely remove support for these environment variables?

**Comment by jack-berg** [comment]
Not sure - we have no precedent for this. The overarching goal here is to have one preferred paradigm to do non-programmatic configuration, with declarative config representing the preference.

Some initial thoughts:

> That we will frezee these environment variables in the specification?

Not sure what you mean by freeze. Perhaps prohibit the expansion of the allowed values? I.e. don't allow expansion of `OTEL_METRICS_EXPORTER` to support a new knew well known value.

>  Do we expect languages to also give some deprecation notice?

I think language SDKs will have to find a way to communicate their plans about this to their users. But I wouldn't imagine a language would deprecate their support for env vars before its declarative config implementation is stable. 

>  Do we want to eventually completely remove support for these environment variables?

I don't think we can remove support from the specification without a major version change. Similarly, I don't think a language SDK would be able to remove support without a major version change. In java, we want to keep publishing a 1.x version for quite some time (I'm tempted to say indefinitely but never say never), and so practically speaking, support for env vars is not going away. Therefore, the deprecation notice is mainly acting a signal to express otel's preference for declarative config, and to indicate that declarative config, not env vars, is where future work to improve configuration will take place.

**Comment by pellared** [comment]
> Not sure what you mean by freeze. Perhaps prohibit the expansion of the allowed values? I.e. don't allow expansion of OTEL_METRICS_EXPORTER to support a new knew well known value.

That is what I had in mind.
